### ISSUE TITLE
About using xatlas to extract uvmap

### ISSUE THREAD

**Issue by jugwangjin** [sbharadwajj/flare#7]
Hi. 
Thank you for great work. 

While I was working on this work, I found out that the resulting uv map is corrupted. 

I fount that although xatlas duplicates some vertices, current work does not apply this on the mesh class.

the resulting uv map is 

![스크린샷 2024-03-05 164209](https://github.com/sbharadwajj/flare/assets/22607015/b89c3721-c40f-44e2-9d7d-0fc58d0b8328)

however, 

if I add two lines in the `xatlas_uvmap` a bit, like
>
    @torch.no_grad()
    def xatlas_uvmap(self):
        import numpy as np
        # Create uvs with xatlas
        v_pos = self.vertices.detach().cpu().numpy()
        t_pos_idx = self.indices.detach().cpu().numpy()
        vmapping, indices, uvs = xatlas.parametrize(v_pos, t_pos_idx)

        # Convert to tensors
        indices_int64 = indices.astype(np.uint64, casting='same_kind').view(np.int64)
        vmapping_int64 = vmapping.astype(np.uint64, casting='same_kind').view(np.int64)
        vmapping = torch.tensor(vmapping_int64, dtype=torch.int64, device=self.device)

        uvs = torch.tensor(uvs, dtype=torch.float32, device=self.device)
        faces = torch.tensor(indices_int64, dtype=torch.int64, device=self.device)

        self._uv_coords = uvs
        self._uv_idx = faces
        self.vmapping = vmapping

        # ADDED LINES
        self.vertices = self.vertices[vmapping]
        self.indices = faces

the resulting UV map goes 

![스크린샷 2024-03-05 164405](https://github.com/sbharadwajj/flare/assets/22607015/1a55b161-d592-423a-9825-ec7a22d880e0)

It may not affect the result, hope it helps. 
especially about the tangent computation, which includes uv coordinates.

**Comment by sbharadwajj** [comment]
Hi,

Thank you so much for your input! Let me recheck. 

As far as I remember, I have verified the UV coordinates in the original code, but a bug might have been introduced when I was cleaning the code for release. Do you see a difference in performance? We don't backpropagate the gradients through this, but the uv maps are used for tangent computation as you mentioned. 

As far as I understand, the step that you mentioned reorders the vertices, right?
I am assuming your UV map is not a face?

**Comment by jugwangjin** [comment]
Thank you for replying. 

As you said, the reordering vetices, can fix this. 

However, when I'm testing on the code, there was no big differences. 
Maybe it can show some robustness on your method. 


ps) the extracted UV map is from the result of xatlas on the face model in the directory 'assets'

**Comment by justin4ai** [comment]
@jugwangjin Hi, thanks for your information. Would you let me know how you did get such an image of uv map in detail? Sadly this model doesn't work in my server, so I have difficulty reverse-engineering after your suggestion.

> Hi. Thank you for great work.
> 
> While I was working on this work, I found out that the resulting uv map is corrupted.
> 
> I fount that although xatlas duplicates some vertices, current work does not apply this on the mesh class.
> 
> the resulting uv map is
> 
> ![스크린샷 2024-03-05 164209](https://github.com/sbharadwajj/flare/assets/22607015/b89c3721-c40f-44e2-9d7d-0fc58d0b8328)
> 
> however,
> 
> if I add two lines in the `xatlas_uvmap` a bit, like
> 
> > 
> 
> ```
> @torch.no_grad()
> def xatlas_uvmap(self):
>     import numpy as np
>     # Create uvs with xatlas
>     v_pos = self.vertices.detach().cpu().numpy()
>     t_pos_idx = self.indices.detach().cpu().numpy()
>     vmapping, indices, uvs = xatlas.parametrize(v_pos, t_pos_idx)
> 
>     # Convert to tensors
>     indices_int64 = indices.astype(np.uint64, casting='same_kind').view(np.int64)
>     vmapping_int64 = vmapping.astype(np.uint64, casting='same_kind').view(np.int64)
>     vmapping = torch.tensor(vmapping_int64, dtype=torch.int64, device=self.device)
> 
>     uvs = torch.tensor(uvs, dtype=torch.float32, device=self.device)
>     faces = torch.tensor(indices_int64, dtype=torch.int64, device=self.device)
> 
>     self._uv_coords = uvs
>     self._uv_idx = faces
>     self.vmapping = vmapping
> 
>     # ADDED LINES
>     self.vertices = self.vertices[vmapping]
>     self.indices = faces
> ```
> 
> the resulting UV map goes
> 
> ![스크린샷 2024-03-05 164405](https://github.com/sbharadwajj/flare/assets/22607015/1a55b161-d592-423a-9825-ec7a22d880e0)
> 
> It may not affect the result, hope it helps. especially about the tangent computation, which includes uv coordinates.

I'd be so glad if you help me out with this!

Best,
Junyeong Ahn

**Comment by jugwangjin** [comment]
> [@jugwangjin](https://github.com/jugwangjin) Hi, thanks for your information. Would you let me know how you did get such an image of uv map in detail? Sadly this model doesn't work in my server, so I have difficulty reverse-engineering after your suggestion.

I captured the UV map on MeshLab, Render -> Show UV Tex Param.


FYI, now I know that the authors' code was absolutely correct.

At that time I was confused about UV parameterization.
What I did was making one-to-one mapping between the all UV vertices and all mesh vertices. 
The results seem the same when we render them, but my code actually splits the mesh surface into patches (same as UV parameterization), instead of a single surface mesh. 
It may affect some geometric regularization process like Laplacian smoothing - Although I did not observed any difference on training results.
